### PR TITLE
ENH: decoupled the codes to use all cache buckets and optimized *zeros* for small chunks of memory. 

### PR DESCRIPTION
**Pull request by blackball**
0. created cache for dimension data, the NCACHE_DIM needs to be discussed. Here I increased it a little because it's unfair that we allowed to cache way more array data than dimension data (1024x7 >> 16x7).   
1. avoided using memory allocation as callback function.  this allows we optimize out most of the *memset()* cost for small memory chunks in *npy_alloc_cache_zero()*.  In the old codes, if `*sz < NBUCKETS*`, it would call *memset* once, no matter the memory was from the cache or newly allocated. 
2. reused all cache buckets. previously, the `data_cache[0]` and `dim_cache[0, 1]` was not used.

### CODE REVIEW

**Comment by charris** [comment]
@juliantaylor Thoughts?
@blackball Your commits will eventually need to be squashed the commit message fixed to conform to the style in `doc/source/dev/gitwash/development_workflow.rst`.

File: numpy/core/src/multiarray/alloc.c [hunk 1]
```diff
@@ -65,8 +63,8 @@ npy_alloc_cache_zero(npy_uintp sz)
 NPY_NO_EXPORT void
 npy_free_cache(void * p, npy_uintp sz)
 {
-    assert(sz > 0);
-    if (p != NULL && sz-1 < NBUCKETS_DATA && datacache[sz-1].available < NCACHE_DATA) {
+    if (p != NULL && sz > 0 && sz-1 < NBUCKETS_DATA &&
+        datacache[sz-1].available < NCACHE_DATA) {
```
Comment (charris) [inline numpy/core/src/multiarray/alloc.c]: Need extra indent here, also in other places.

**Comment by charris** [comment]
C style also needs attention, see `doc/C_STYLE_GUIDE.rst.txt`.

**Comment by blackball** [comment]
@charris I have formatted the codes according to the style guides and created a branch(blackball:improve_alloc) for it. I will close this request, please refer to the new one.

**Comment by eric-wieser** [comment]
@blackball: In future there's no need to do that. You can always just force-push over the same PR.

**Pull request closed without merging**